[MM-18898] Stringify plugin.Log* parameters

cinlloc opened the pull request:
#### Summary
Convert object parameters of `plugin.Log*` RPC methods to their string version to avoid gob conversion error. More info on commit 4955c63 message. 

#### Ticket Link
Implements #12381

#### Comments
- Tested with a plugin in `app/plugin_api_test.go`, "gob" error is not visible anymore. Not sure it worth having it in codebase as I wasn't able to fully automate it (visual check of the logs)
- Added a `stringutils` because haven't found better place

lieut-data commented on utils/stringutils/stringutils_test.go:
```diff
@@ -0,0 +1,92 @@
+	assert.Empty(t, strings)
+}
+
+func TestStringifyEmptyShouldReturnEmpty(t *testing.T) {
```
nit: would it make sense to have a single `TestStringify` with subtests instead (similarly below)?

cinlloc replied:
Just to be sure, are you talking about something like that? https://ieftimov.com/post/testing-in-go-subtests/
One drawback I can see doing this is that subtests force to use only `assert.Equal` method, but I can get a try if you feel so.

cinlloc replied:
I propose a new patch using subtests. I may slightly prefer the previous version which is more TDD approach, one behavior / one function, but as you wish ;-)

lieut-data replied:
The approach you took is definitely valid, and I have no objections.

One other approach is just to think of subtests as a means of grouping, e.g.:
```
func TestStringify(t *testing.T) {
    t.Run("nil should return empty", func(t *testing.T) {
        // ...
    })
    t.Run("empty should return empty", func(t *testing.T) {
        // ...
    })

lieut-data replied:
I prefer the above when the amount of duplication is low (e.g. in this case, where the SUT is basically just two lines.) The other approach is great when there's a lot of complex setup and teardown that you'd like to avoid duplicating.

ali-farooq0 commented on app/plugin_api_test.go:
```diff
@@ -1440,6 +1440,17 @@ func TestPluginAPIGetUnsanitizedConfig(t *testing.T) {
+	pluginCode, err := ioutil.ReadFile(filepath.Join(path, "plugin_using_log_api.go"))
+	assert.NoError(t, err)
+	setupPluginApiTest(t, string(pluginCode),
+		`{"id": "com.mattermost.sample", "server": {"executable": "backend.exe"}, "settings_schema": {"settings": []}}`, pluginID, th.App)
```
really sad that we can't really dependency inject the logger and verify if the call worked correctly 😢

cinlloc replied:
I agree! There may be a solution but I haven't found it!

prapti commented:
@lieut-data can the [ticket](https://mattermost.atlassian.net/browse/MM-18898) for this PR be updated with some QA test steps? That'd be helpful.

lieut-data commented:
@prapti, done!

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-server-pr-12700.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

cinlloc commented:
Working on another PR, it occurred to me that I missed API documentation update of the Log* methods (very sorry for this!). This line should be removed from each Log* method: 
`keyValuePairs should be primitive go types or other values that can be encoded by encoding/gob`
e.g. https://github.com/cinlloc/mattermost-server/blob/6f4f06f8d4a586bf3deaabb8eaad9dde3fbe7465/plugin/api.go#L636
As the QA is ongoing, what is the best way to handle this?
- Push a commit with the doc fix and don't do the QA again, because it's only comment so no functional impact?
- Fix in another PR?
- Push a commit on this PR with the fix and redo the QA? :-(
- Keep this line as is, not a big deal

cinlloc commented:
@lieut-data good catch indeed! I implemented your suggestion and added new tests, especially about primitive types, as they are now stringified before RPC call.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `27410d6b009b647e87d5b1d10ed12beef7fe2a51`.

Access here: https://mattermost-server-pr-12700.test.mattermost.cloud

mattermod commented:
Test server destroyed

lieut-data commented on plugin/stringifier_test.go:
```diff
@@ -4,41 +4,56 @@
-		t.Run(c.name, func(t *testing.T) {
-			strings := stringify(c.objects)
-			assert.Equal(t, c.strings, strings)
+	t.Run("NilShouldReturnEmpty", func(t *testing.T) {
```
0/5, but it's fine to have spaces in the subtest names

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-server-pr-12700.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mattermod commented:
Test server destroyed
